lib/backend/fs/impl.go

```diff
@@ -0,0 +1,310 @@
+	if err := os.MkdirAll(bk.LocksDir, defaultDirMode); err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+	return bk, nil
+}
+
+// GetKeys returns a list of keys for a given path
+func (bk *Backend) GetKeys(bucket []string) ([]string, error) {
+	files, err := ioutil.ReadDir(path.Join(bk.RootDir, path.Join(bucket...)))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return []string{}, nil
+		}
+		return nil, trace.ConvertSystemError(err)
+	}
+	// enumerate all directory entries and select only non-hidden files
+	retval := make([]string, 0)
+	for _, fi := range files {
+		name := fi.Name()
+		if !fi.IsDir() && name[0] != '.' {
+			retval = append(retval, name)
+		}
+	}
+	return retval, nil
+}
+
+// CreateVal creates value with a given TTL and key in the bucket
+// if the value already exists, returns AlreadyExistsError
+func (bk *Backend) CreateVal(bucket []string, key string, val []byte, ttl time.Duration) error {
+	log.Debugf("fs.CreateVal(%s/%s) '%v'", strings.Join(bucket, "/"), key, string(val))
+	// do not allow keys that start with a dot
+	if key[0] == '.' {
+		return trace.BadParameter("Invalid key: '%s'. Key names cannot start with '.'", key)
+	}
+	// create the directory:
+	dirPath := path.Join(bk.RootDir, path.Join(bucket...))
+	err := os.MkdirAll(dirPath, defaultDirMode)
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	// create the file (AKA "key"):
+	filename := path.Join(dirPath, key)
+	f, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_EXCL, defaultFileMode)
+	if err != nil {
+		if os.IsExist(err) {
+			return trace.AlreadyExists("%s/%s already exists", dirPath, key)
+		}
+		return trace.ConvertSystemError(err)
+	}
+	defer f.Close()
+	n, err := f.Write(val)
+	if err == nil && n < len(val) {
+		return trace.Wrap(io.ErrShortWrite)
+	}
+	bk.applyTTL(dirPath, key, ttl)
+	return nil
+}
+
+// UpsertVal updates or inserts value with a given TTL into a bucket
+// ForeverTTL for no TTL
+func (bk *Backend) UpsertVal(bucket []string, key string, val []byte, ttl time.Duration) error {
+	log.Debugf("fs.UpsertVal(%s/%s) '%v'", strings.Join(bucket, "/"), key, string(val))
+	// create the directory:
+	dirPath := path.Join(bk.RootDir, path.Join(bucket...))
+	err := os.MkdirAll(dirPath, defaultDirMode)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// create the (or overwrite existing) file (AKA "key"):
+	return ioutil.WriteFile(path.Join(dirPath, key), val, defaultFileMode)
+}
+
+// GetVal return a value for a given key in the bucket
+func (bk *Backend) GetVal(bucket []string, key string) ([]byte, error) {
+	log.Debugf("fs.GetVal(%s/%s)", strings.Join(bucket, "/"), key)
+	dirPath := path.Join(path.Join(bk.RootDir, path.Join(bucket...)))
+	expired, err := bk.checkTTL(dirPath, key)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if expired {
+		bk.DeleteKey(bucket, key)
+		return nil, trace.NotFound("key '%s' is not found", key)
+	}
+	return ioutil.ReadFile(path.Join(dirPath, key))
```

Comment:
I would suggest using `https://github.com/gravitational/teleport/blob/master/lib/utils/fs.go#L79` as it's a bit more detailed in it's errors